### ISSUE TITLE
IDEFICS Processor will not respect `padding_side="right"` when `padding="longest"`

### ISSUE THREAD

**Issue by xhluca** [huggingface/transformers#26354]
### System Info

- `transformers` version: 4.33.2
- Platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.34
- Python version: 3.8.16
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.3
- Accelerate version: 0.22.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@VictorSanh 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoProcessor

processor = AutoProcessor.from_pretrained("HuggingFaceM4/idefics-9b", padding_side="right")

sents = [["hello world"], [" this is a longer sentence to testing padding"]]

# This is the correct behavior:
a = processor(sents, padding="max_length", truncation=True, max_length=20)
print(processor.tokenizer.decode(a['input_ids'][0]))
# => <s> hello world<unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk>

# This is the incorrect behavior:
b = processor(sents, padding="longest", truncation=True, max_length=30)

print(processor.tokenizer.decode(b['input_ids'][0]))
# => <unk><unk><unk><unk><unk><unk><s> hello world
```

### Expected behavior

It should pad on right side

**Comment by shauray8** [comment]
@ArthurZucker Not sure, but maybe this is because idefics processor forces the tokenizer to take the prompts one by one and for `padding="longest"` to work it has to take all the prompts at once.

https://github.com/huggingface/transformers/blob/main/src/transformers/models/idefics/processing_idefics.py#L285-L333

*a potential fix would be to get the text encoding part of it out of the loop, it should be a fairly quick fix (if there are no border conditions following this)*

**Comment by LysandreJik** [comment]
This indeed looks like a bug cc @leot13 @ArthurZucker

**Comment by VictorSanh** [comment]
This is definitely a bug. with a quick dive into the code, i think what @shauray8  makes sense.
It's mostly about moving things around: unindenting the block 324-330, let the tokenizer handle the padding logic (block 346-352).
(i might be wrong, just having a quick look)
If you feel like it, feel free to open a PR @shauray8 / @xhluca

**Comment by shauray8** [comment]
On it!
